aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h

```diff
@@ -1067,6 +1067,21 @@ CONVERT_VECTORIZED_INIT(BFloat16, bfloat16);
 CONVERT_VECTORIZED_INIT(Half, half);
 
 #else // defined(CPU_CAPABILITY_AVX2)
+template <>
+inline void convert(const float* src, BFloat16* dst, int64_t n) {
+  // TODO: use simd intrinsics on the platforms that support bfloat16
+  for (int64_t i = 0; i < n; i++) {
+    dst[i] = c10::convert<BFloat16>(src[i]);
+  }
+}
+
+template <>
+inline void convert(const BFloat16* src, float* dst, int64_t n) {
+  // TODO: use simd intrinsics on the platforms that support bfloat16
+  for (int64_t i = 0; i < n; i++) {
+    dst[i] = c10::convert<float>(src[i]);
+  }
+}
```

Comment:
Since this is also working on scalars with a for-loop, curious what is the difference from the vanilla unspecialized implementation from vec_base.h?
```c++
template <typename src_T, typename dst_T>
inline void convert(const src_T *src, dst_T *dst, int64_t n) {
#ifndef _MSC_VER
# pragma unroll
#endif
  for (C10_UNUSED const auto i : c10::irange(n)) {
    *dst = c10::convert<dst_T>(c10::load(src));
    src++;
    dst++;
  }
}
```

Reply:
Good question. I captured some profiling from the default implementation. That particular convert() function takes 6.2% of CPU time. And the integer_iterator's operator, which is the callee from convert(), takes 3.7% of the CPU time. For convert(), more than half of the time is spent on iterator. This could be saved with this PR.  
 
<img width="1183" alt="before" src="https://github.com/pytorch/pytorch/assets/18431130/98e6b9cb-1f40-459c-a922-5b67732d0cda">

Reply:
I'm also wondering why `irange` would incur such significant overhead. The code looks trivial and a decent compiler should be able to optimize away the abstraction overhead? https://github.com/pytorch/pytorch/blob/main/c10/util/irange.h#L21 

A trivial experiment shows that compiler can optimize away `irange`: https://godbolt.org/z/MaqqbjEcM 

Is there some downstream effect like `irange` interfering with SEVC which in turns blocks other loop optimization?

Reply:
> I'm also curious about this! I introduced itange everywhere because folks kept messing up the data type on for loops thereby artificially restricting data sizes. But if it has significant perf implications, we may have to rethink it. It's also interesting because C++20 will provide similar functionality out of the box.
> 
> @jgong5 : we'd keep irange for consistency. There were hundreds or even thousands of badly written for loops that it fixed. I'd prefer we didn't backslide on that.

OK. But can we add some comments on the new code to note this background?

Reply:
Those are good questions. Sorry for the late reply. I will check the assembly to see what's going on there.